noir-projects/aztec-nr/aztec/src/state_vars/shared_mutable/shared_mutable_values.nr

```diff
@@ -1,55 +1,119 @@
-    scheduled_value_change::ScheduledValueChange,
+use crate::{
+    state_vars::shared_mutable::{
+        scheduled_delay_change::ScheduledDelayChange, scheduled_value_change::ScheduledValueChange,
+    },
+    utils::array,
 };
-use crate::utils::array;
-use dep::protocol_types::{traits::Packable, utils::arrays::array_concat};
+use dep::protocol_types::traits::Packable;
 use std::meta::derive;
 
+mod test;
+
 /// SharedMutableValues is just a wrapper around ScheduledValueChange and ScheduledDelayChange that then allows us
 /// to wrap both of these values in WithHash. WithHash allows for efficient read of values in private.
 ///
 /// Note that the WithHash optimization does not work in public (due to there being no unconstrained). But we also want
-/// to be able to read the values efficiently in public and we want to be able to read each value separately. For that
-/// reason we expose `get_delay_change_storage_slot` and `get_value_change_storage_slot` which point to the correct
-/// location in the storage. This is "hacky" as we pack and store the values together but there is no way around it.
+/// to be able to read the values efficiently in public and we want to be able to read each value separately. Reading
+/// the values separately is tricky because ScheduledValueChange and ScheduledDelayChange are packed together (sdc and
+/// svc.block_of_change are stored in the same slot). For that reason we expose `unpack_value_change` and
+/// `unpack_delay_change` functions that can be used to extract the values from the packed representation. This
+/// is "hacky" but there is no way around it.
 #[derive(Eq)]
 pub(crate) struct SharedMutableValues<T, let INITIAL_DELAY: u32> {
     svc: ScheduledValueChange<T>,
     sdc: ScheduledDelayChange<INITIAL_DELAY>,
 }
 
-impl<T, let INITIAL_DELAY: u32> SharedMutableValues<T, INITIAL_DELAY> {
+impl<T, let INITIAL_DELAY: u32, let N: u32> SharedMutableValues<T, INITIAL_DELAY>
+where
+    T: Packable<N>,
+{
     pub(crate) fn new(
         svc: ScheduledValueChange<T>,
         sdc: ScheduledDelayChange<INITIAL_DELAY>,
     ) -> Self {
         SharedMutableValues { svc, sdc }
     }
 
-    pub fn get_delay_change_storage_slot(shared_mutable_storage_slot: Field) -> Field {
-        shared_mutable_storage_slot
+    pub(crate) fn unpack_value_change(packed: [Field; 2 * N + 1]) -> ScheduledValueChange<T> {
+        let svc_pre_packed = array::subarray(packed, 1);
+        let svc_post_packed = array::subarray(packed, N + 1);
+        ScheduledValueChange::new(
+            T::unpack(svc_pre_packed),
+            T::unpack(svc_post_packed),
+            packed[0] as u32,
+        )
     }
 
-    pub fn get_value_change_storage_slot(shared_mutable_storage_slot: Field) -> Field {
-        shared_mutable_storage_slot + (SCHEDULED_DELAY_CHANGE_PCKD_LEN as Field)
+    pub(crate) fn unpack_delay_change(packed: Field) -> ScheduledDelayChange<INITIAL_DELAY> {
```

Comment:
Using modulo operation in this func was not much of an option since the packed value does not fit in `U128` and Field does not support it. For this reason I went with the "tmp value approach".